### ISSUE TITLE
SL Training Schedule

### ISSUE THREAD

**Issue by wrongu** [Rochester-NRT/RocAlphaGo#148]
Add a configurable learning rate schedule for supervised learning.

**Comment by MaMiFreak** [comment]
looks like the easiest way to implement is to use a LearningRateScheduler but then it only possible to decrease the learning rate at an epochs end/start.
say:
--learning-rate 0.03
--epoch-length 5,000,000
--decay 0.5
--drop_epoch 40   -> better name would be welcome

learning starts with lr = 0.03 
after 40 epochs lr drops to 0.015   ( step 80,000,000 = drop_epoch * epoch-lenght )
after 80 epochs lr drops to 0.0075 ( step 160,000,000 = drop_epoch * epoch-lenght )

so user has to set epoch-lenght, and drop epoch appropriate to the step lr should drop
epoch-lenght 20,000,000 and drop_epoch 4 would drop during the same step

when drop_epoch is not set decay works as it does now

i'll have a look at a way to drop at a certain step but so far i haven't found a callback that is being called every step

**Comment by MaMiFreak** [comment]
alphago paper mentioning halving after 80,000,000 steps.
alphago step = 16 positions ( 1 batch )
rocalphago step = 1 position

i am not suggesting changing it but probably worth mentioning somewhere, not sure where though...

**Comment by wrongu** [comment]
Maybe `--drop-epoch` could be `--decay-period` or `--decay-every`? `LearningRateScheduler` is exactly what we need. Something like this:

```python
def lr_schedule(epoch):
    n_decay = int(epoch / args.decay_every)
    return args.lr * args.decay ** n_decay
scheduler_callback = LearningRateScheduler(lr_schedule)
...
model.fit_generator(
    generator=train_data_generator,
    samples_per_epoch=samples_per_epoch,
    nb_epoch=args.epochs,
    callbacks=[checkpointer, meta_writer, scheduler_callback],
    validation_data=val_data_generator,
    nb_val_samples=n_val_data)
```

And since callbacks will (soon) have an adjusted `initial_epoch`, it will work with restarting as well.

Good point about step size. Maybe could be mentioned in the documentation of SL when you run with `-h`

**Comment by MaMiFreak** [comment]
i found `on_batch_begin` and `on_batch_end` options in keras
no calculation required by user and difference in step is removed

**Comment by MaMiFreak** [comment]
https://github.com/MaMiFreak/RocAlphaGo/tree/MaMiFreak-symmetries

not ready! but
- decay after x batches works, use "--decay-every" (theano backend verfication required)
- correct weight file names after restart
- correct batch number is used after restart

**Comment by MaMiFreak** [comment]
option is added to supervised trainer
